### DISCUSSION TITLE
Feature request: Double Metaphone string encoder

### DISCUSSION THREAD

**Discussion by mamonu** [duckdb/duckdb#5148] [General]
Text function feature request: a Double Metaphone string encoder

Double Metaphone is a phonetic algorithm for indexing words by their English pronunciation. It fundamentally improves on previous related  algorithms by using information about variations and inconsistencies in English spelling and pronunciation to produce a more accurate encoding, which does a better job of matching words and names which sound similar. 
For example Smith, Smyth, Smithe, Smythe all encode to SM0.

It is one of the important functions used in record linkage. Its functionality sits among other related functions such as Levenstein distance or the newly implemented Jaro-Winkler distance (https://github.com/duckdb/duckdb/pull/4225) only in this case its not a distance  (  d(x,y) = z ) but a scalar function ( f(x)= y) where there is one (string) input and one string (double metaphone mapping) output. 

Adding this function to duckdb to sit with the rest of the linkage related functions (Jaccard/ Levenstein /JaroWinkler) will be very useful.

Some background and a Java battle tested implementation below:

[1] [wikipedia article](https://en.wikipedia.org/wiki/Metaphone)

[2] [Java implementation on Apache Commons](https://commons.apache.org/proper/commons-codec/apidocs/src-html/org/apache/commons/codec/language/DoubleMetaphone.html)

**Comment by chris-park** [comment]
I also raised this as a discussion point a few weeks ago: https://github.com/duckdb/duckdb/discussions/4818.

**Comment by Maxxen** [comment]
If @Mytherin would find this suitable to be in core I could maybe get around to implementing this. Although I don't know how the license would work if we ported it over from java.
